input.go

```diff
@@ -0,0 +1,29 @@
+package shaping
+
+import "image/color"
+
+type Input interface {
+	// Text returns the characters to be shaped.
+	Text() []rune
+	// Color returns the color to render the text in.
+	Color() color.Color
+
+	// TODO should these be wrapped in an interface, eg. Font() image/font?
```

Comment:
I do think we should wrap the font in an interface, and I saw the discussion around this in slack. I don't have strong opinions about what that interface should look like, but it does need to be flexible to a wide variety of kinds of font (bitmap, vector, many formats, fonts with fallback fonts, etc ...)

Reply:
Yes. Also as this is on the "input" and not "state" there can be only 1 font attached, so the use of face and family are probably not required, this would be a single font instance and the attributes that led to it being loaded are not likely to help the shaping as the bold/italic is set in the font/face itself.

Reply:
I agree, this seems like a better approach; first a font is loaded into the shaper library, then each input references the font by family name (eg. Noto Sans) and includes attributes such as `Bold() bool`, `Italic() bool`, `Monospace() bool`, and `ScriptVariant() ScriptVariant` would be used by the shaper to select the right member of the font family (eg. Noto Sans Bold).

Reply:
That's not really what I was saying, why would we want to load all of these variants of a font when the call to shape can only use 1 style?

Reply:
I also think it is better to select the font from style attributes in a preliminary step, not covered by the shaper itself. In the linux stack, this is usually done by fontconfig : pango collects style attributes, pass them to fontconfig, which returns a font file, which is then passed to Harfbuzz. As it was already mentioned, Harfbuzz itself doesn't known if the font was bold or italic. This matches the fact that two styles (regular or bold say) are usually coming from two different font files (or two different fonts in a collection).

Reply:
Ah I see, so instead of loading a whole font family, only a particular variant is loaded and referenced by the input to the shaper.

So most of the attributes (Bold, Italic, Sub/Superscript, Monospace) can be removed from the input and moved to a library called earlier in the pipeline?

Reply:
Yes, quite possibly the toolkit themselves will want to control the source of font data.